backend/corpora/dataset_processing/remaster_cxg.py

```diff
@@ -0,0 +1,250 @@
+    logger.info("Reading array to determine sparsity...")
+    with tiledb.scope_ctx(ctx):
+        with tiledb.open(f"{cxg}/{source_array}", "r") as X:
+            nnz = 0
+            i, chunk = 0, 200_000
+            while i < X.shape[0]:
+                logger.info(f"Chunk {i}")
+                dat = X.query(dims=[])[i : i + chunk]
+
+                if X.schema.sparse:
+                    nnz += len(dat[""])
+                else:
+                    nnz += np.count_nonzero(dat)
+                i += chunk
+
+            size = X.shape[0] * X.shape[1]
+            logger.info(f"sparsity={1.0 - nnz / size}, nnz={nnz}, extent=({X.shape[0]}, {X.shape[1]})")
+            if 1.0 - nnz / size >= sparse_threshold / 100.0:
+                return "sparse"
+
+            return "dense"
+
+
+def create_new_X(cxg, target_array, kind, compression, X_extent, old_schema, cell_order, tile_order, capacity):
+    is_sparse = kind == "sparse"
+    logger.info(f"create_new_X: {is_sparse}")
+    attr_filters = tiledb.FilterList([tiledb.ZstdFilter(level=compression)])
+    dim_filters = tiledb.FilterList([tiledb.ByteShuffleFilter(), tiledb.ZstdFilter(level=compression)])
+    old_dims = [old_schema.domain.dim(d) for d in range(old_schema.domain.ndim)]
+    old_attr = old_schema.attr(0)
+
+    tiledb.Array.create(
+        f"{cxg}/{target_array}",
+        tiledb.ArraySchema(
+            domain=tiledb.Domain(
+                [
+                    tiledb.Dim(
+                        name="obs",
+                        domain=old_dims[0].domain,
+                        tile=min(old_dims[0].domain[1], X_extent[0]),
+                        dtype=old_dims[0].dtype,
+                        filters=dim_filters,
+                    ),
+                    tiledb.Dim(
+                        name="var",
+                        domain=old_dims[1].domain,
+                        tile=min(old_dims[1].domain[1], X_extent[1]),
+                        dtype=old_dims[1].dtype,
+                        filters=dim_filters,
+                    ),
+                ]
+            ),
+            sparse=is_sparse,
+            allows_duplicates=True if is_sparse else False,
+            attrs=[tiledb.Attr(name="", dtype=old_attr.dtype, filters=attr_filters)],
+            cell_order=f"{cell_order}-major",
+            tile_order=f"{tile_order}-major",
+            capacity=capacity if is_sparse else 0,
+        ),
+    )
+
+
+def create_ctx(config: dict = {}) -> tiledb.Ctx:
+    cfg = tiledb.Config(config)
+    ctx = tiledb.Ctx(config=cfg)
+    return ctx
+
+
+def frac_mem(f):
+    mem_size = psutil.virtual_memory().total
+    return int(f * mem_size) // (1024**2) * (1024**2)
+
+
+def fast_config(config_overrides: dict = {}) -> dict:
+    # consolidation buffer heuristic to prevent thrashing: total_mem/io_concurrency_level, rounded to GB
+    io_concurrency_level = int(tiledb.Config()["sm.io_concurrency_level"])
+    consolidation_buffer_size = (
```

Comment:
this was originally written for very large memory instances.  I don't know what environment it will run in, but if any have small available memory (eg, < 1GB), I suggest just using the default, or bumping the 10% to something larger.